### ISSUE TITLE
[Feature]: Infer form validation from HTML5 attributes?

### ISSUE THREAD

**Issue by LouisStAmour** [remix-run/remix#539]
### What is the new or updated feature that you are suggesting?

When writing a form using the uppercase Form tag, you rely on HTML semantics for submitting the form by default, and need very little code to do so.

But then when you have to validate inputs - extremely common and very good advice - you have to write your own function. Really?

If we can infer the data model being sent from the inputs, we could infer the validation required simply by putting HTML5 attributes on the relevant input tags. Then just as the browser will block submission when inputs are invalid in real-time on the client (and we can attach CSS to this), so too can the server which created the HTML for the form look at the form’s HTML to infer what the validation should be - which fields are required, min and max length, regex matching, and so on. By default, https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation could be validated server-side, and if such validation isn’t desired it could be turned off with an attribute on the Form tag, perhaps.

### Why should this feature be included?

It’s not acceptable to compromise on usability by delivering sub-par real-time input validation. And if a rule can be expressed within the confines of html client side validation, then for simplicity’s sake, we shouldn’t reinvent the wheel, we should use the attributes in HTML to perform the validation. If the data model lives in the form HTML, then the validation should too, if simple enough.

**Comment by sean-brydon** [comment]
Instead of writing your own functions you could rely on something like ZOD do to the work for you. With the power of typescript you can also make this DevX quite nice. See below:

Create function to parse form from and comply with zod schema
```tsx
import { ZodObject, ZodRawShape } from "zod";

export default function useFormData<T extends ZodRawShape>(
  form: FormData,
  inputSchema: ZodObject<T>
) {
  const keys = Object.keys(inputSchema.shape);
  let returnObject: Record<string, any> = {};
  keys.forEach((item) => {
    returnObject[item] = form.get(item);
  });

  return inputSchema.parse(returnObject);
}

```

Use this function as follows `routes/login.ts`
```tsx

const LoginObjectSchema = z.object({
  username: z.string(),
  password: z.string().min(5),
  redirectTo: z.string().optional(),
});

// Create a response model
export const action: ActionFunction = async ({
  request,
}): Promise<Response | any> => {
  const form = await request.formData();
  const { username, password, redirectTo } = getFormData(
    form,
    LoginObjectSchema
  );
 // ... Not needed in example
};

export default function Login() {
  const [searchParams] = useSearchParams();
  return (
    <Form action="/login" method="post">
      <input
        type="hidden"
        name="redirectTo"
        value={searchParams.get("redirectTo") ?? undefined}
      />
      <div className="">
        <label htmlFor="username-input"></label>
        <input type="text" name="username" required />
      </div>
      <div className="">
        <label htmlFor="password-input"></label>
        <input type="password" name="password" required />
      </div>
      <button type="submit">Log In</button>
    </Form>
  );
}

```
 Note: This solution may not be exactly what you were asking for but it is defo a better approach than the example given in the docs

**Comment by LouisStAmour** [comment]
Interesting. But yeah, my ask is more like being able to say:

```ts
export default function Login() {
  return (
    <Form action="/login" method="post">
        <input type="text" name="username" required />
        <input type="password" name="password" required minlength="5" />
      <button type="submit">Log In</button>
    </Form>
  );
}
```

And have the following get generated automatically from the above function, perhaps with something like `const LoginObjectSchema = formToZod(Login);` or that at some later step, the `Form` can look at its rendered children to infer the data model or schema. For example, maybe Form could use https://github.com/grrowl/react-keyed-flatten-children or similar to see what input children it has, in a similar way to how React-Router can use children to know what routes have been provided to the router.

```ts
// What I don't want to have to write:
const LoginObjectSchema = z.object({
  username: z.string(),
  password: z.string().min(5),
  redirectTo: z.string().optional(),
});
```

**Comment by sean-brydon** [comment]
My guess it there would need to be some serverside magic happening here as if you were to generate it on the client side anyone could theoretically change your input validation.  I completely understand why you'd want what youre asking for - but looking from the outside in it could potentially be a massive pain to implement. 

If you have any ideas for the approach im happy to take a look.

**Comment by LouisStAmour** [comment]
There's already server-side magic happening for routing... if we can process JSX server-side for:

```
<Route path="/" element={<App />}>
  <Route path="sales" element={<Sales />}>
    <Route path="invoices" element={<Invoices />}>
      <Route path=":invoice" element={<Invoice />} />
    </Route>
  </Route>
</Route>
```
E.g. this would be an awesome way to write a login form:

```
const LoginForm = ({fields}) => (
<Form fields={fields} onValidSubmit={loginFunction}>
  <TextInput name="username" required {...fields.username}>Username</TextInput>
  <PasswordInput name="password" required minLength={5} {...fields.password}>Password</TextInput>
</Form>
);
```

Where `fields` might be something like the current state of the form
```js
{ "username": {
  value: "current value from a form submit with errors...",
  errors: ["Error 1", "Some other error message"]
}, "password": { 
  value: "", 
  errors: ["cannot be left blank"]
} }
```

**Comment by sean-brydon** [comment]
> ```js
> FormBuilder({
>   username: {label: "Username", required: true},
>   password: {label: "Password", required: true, minLength: 5,
>   buttonText: "Login"
> });
> ```
> 
> It's too abstract from the HTML it generates later, after all. At that point, might as well build a model object because that's what you're describing, with the exception of the labels.

Yeah i feel like if we're at this point - using zod or any sort of schema validation may as well be implemeneted. :/ Sorry for taking this issue slightly off topic - Just wanted to share my approach to solving a simular sort of issue
